File: algorithms/hard/cherry_pickup.ex
defmodule Solution do
  @min -300

  @spec cherry_pickup(grid :: [[integer]]) :: integer
  def cherry_pickup(grid) do
    n = Enum.count(grid) - 1

    grid =
      grid
      |> Enum.with_index()
      |> Enum.reduce(%{}, fn {row, i}, acc ->
        row
        |> Enum.with_index()
        |> Enum.reduce(acc, fn {n, j}, acc ->
          Map.put(acc, {i, j}, n)
        end)
      end)

    dp = %{{0, 0, 0} => grid[{0, 0}]}

    for x1 <- 0..n, y1 <- 0..n, x2 <- 0..n, reduce: dp do
      dp ->
        y2 = x1 + y1 - x2

        if Map.get(dp, {x1, y1, x2}, @min) > 0 || y2 < 0 || y2 > n || grid[{x1, y1}] == -1 ||
             grid[{x2, y2}] == -1 do
          dp
        else
          pre =
            Enum.max([
              Map.get(dp, {x1 - 1, y1, x2 - 1}, @min),
              Map.get(dp, {x1 - 1, y1, x2}, @min),
              Map.get(dp, {x1, y1 - 1, x2 - 1}, @min),
              Map.get(dp, {x1, y1 - 1, x2}, @min)
            ])

          if pre < 0 do
            dp
          else
            cherries =
              (x1 == x2 && grid[{x1, y1}]) ||
                grid[{x1, y1}] + grid[{x2, y2}]

            Map.put(dp, {x1, y1, x2}, pre + cherries)
          end
        end
    end
    |> Map.get({n, n, n}, 0)
  end
end

File: algorithms/hard/cherry_pickup.rs
impl Solution {
    pub fn cherry_pickup(grid: Vec<Vec<i32>>) -> i32 {
        let n = grid.len();
        let mut dp = vec![vec![vec![i32::MIN; n+1]; n+1]; n+1];
        dp[1][1][1] = grid[0][0];
        for x1 in 1..=n {
            for y1 in 1..=n {
                for x2 in 1..=n {
                    let y2 = x1 + y1 - x2;
                    if dp[x1][y1][x2] > 0 || y2 < 1 || y2 > n || grid[x1-1][y1-1] == -1 || grid[x2-1][y2-1] == -1 {
                        continue;
                    }
                    let pre = vec![
                        dp[x1-1][y1][x2-1],
                        dp[x1-1][y1][x2],
                        dp[x1][y1-1][x2-1],
                        dp[x1][y1-1][x2],
                    ].into_iter().max().unwrap();
                    if pre < 0 {
                        continue;
                    }
                    dp[x1][y1][x2] = pre + grid[x1-1][y1-1];
                    if x1 != x2 {
                        dp[x1][y1][x2] += grid[x2-1][y2-1];
                    }
                }
            }
        }
        dp[n][n][n].max(0)
    }
}
